tests/0098-consumer-txn.cpp

```diff
@@ -0,0 +1,892 @@
+                       rapidjson::GetParseError_En(d.GetParseError()) <<
+                       " at " << d.GetErrorOffset());
+
+          rd_snprintf(path, sizeof(path),
+                      "/topics/%s/partitions/0",
+                      topic.c_str());
+
+          rapidjson::Pointer jpath((const char *)path);
+          rapidjson::Value *pp = rapidjson::GetValueByPointer(d, jpath);
+          if (pp == NULL)
+            return;
+
+          test_assert(pp->HasMember("hi_offset"),
+                      "hi_offset not found in stats");
+          test_assert(pp->HasMember("ls_offset"),
+                      "ls_offset not found in stats");
+
+          partition_0_hi_offset = (*pp)["hi_offset"].GetInt();
+          partition_0_ls_offset = (*pp)["ls_offset"].GetInt();
+        }
+        break;
+
+      default:
+        break;
+    }
+  }
+};
+
+bool TestEventCb::should_capture_stats;
+bool TestEventCb::has_captured_stats;
+int64_t TestEventCb::partition_0_hi_offset;
+int64_t TestEventCb::partition_0_ls_offset;
+std::string TestEventCb::topic;
+
+static TestEventCb ex_event_cb;
+
+
+static void execute_java_produce_cli(std::string &bootstrapServers,
+                                     std::string &topic, std::string cmd) {
+  const char *argv[] = {
+    bootstrapServers.c_str(),
+    topic.c_str(),
+    cmd.c_str(),
+    NULL
+  };
+
+  int pid = test_run_java("TransactionProducerCli", argv);
+  test_waitpid(pid);
+}
+
+static std::vector<RdKafka::Message *> consume_messages(
+                                          RdKafka::KafkaConsumer *c, 
+                                          std::string topic,
+                                          int partition) {
+  RdKafka::ErrorCode err;
+
+  /* Assign partitions */
+  std::vector<RdKafka::TopicPartition*> parts;
+  parts.push_back(RdKafka::TopicPartition::create(topic, partition));
+  if ((err = c->assign(parts)))
+    Test::Fail("assign failed: " + RdKafka::err2str(err));
+  RdKafka::TopicPartition::destroy(parts);
+
+  Test::Say("Consuming from topic " + topic + "\n");  
+  std::vector<RdKafka::Message *> result = std::vector<RdKafka::Message *>();
+
+  while (true) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    switch (msg->err())
+    {
+      case RdKafka::ERR__TIMED_OUT:
+        continue;
+      case RdKafka::ERR__PARTITION_EOF:
+        break;
+      case RdKafka::ERR_NO_ERROR:
+        result.push_back(msg);
+        continue;
+      default:
+        Test::Fail("Error consuming from topic " + 
+                   topic + ": " + msg->errstr());
+        break;
+    }
+    break;
+  }
+
+  Test::Say("Read all messages from topic: " + topic + "\n");
+
+  TestEventCb::should_capture_stats = true;
+
+  /* rely on the test timeout to prevent an infinite loop in
+   * the (unlikely) event that the statistics callback isn't
+   * called. */
+  while (true) {
+    c->consume(tmout_multip(500));
```

Comment:
this is leaking memory if a message or consumer error (which is a Message with err set) is returned.

Reply:
there are also memory leaks above.